apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_state.erl

```diff
@@ -654,19 +1169,77 @@ pmap_open(Table, SessionId) ->
         cache = Clean,
         dirty = #{}
     }.
+%% END ifdef(STORE_STATE_IN_DS).
+-endif.
 
 -spec pmap_get(K, pmap(K, V)) -> V | undefined.
+-ifdef(STORE_STATE_IN_DS).
+pmap_get(K, #pmap{table = Table, key_mapping = KeyMapping, cache = Cache}) when
+    is_map_key(K, KeyMapping)
+->
+    IntK = maps:get(K, KeyMapping),
+    cache_get(Table, IntK, Cache);
+pmap_get(_K, _Pmap) ->
+    undefined.
+%% ELSE ifdef(STORE_STATE_IN_DS).
+-else.
 pmap_get(K, #pmap{table = Table, cache = Cache}) ->
     cache_get(Table, K, Cache).
+%% END ifdef(STORE_STATE_IN_DS).
+-endif.
 
 -spec pmap_put(K, V, pmap(K, V)) -> pmap(K, V).
+-ifdef(STORE_STATE_IN_DS).
+pmap_put(
+    K, V, Pmap = #pmap{table = Table, key_mapping = KeyMapping0, dirty = Dirty, cache = Cache}
+) ->
+    {IntK, KeyMapping} = get_or_gen_internal_key(K, KeyMapping0, Table, Cache),
+    Pmap#pmap{
+        key_mapping = KeyMapping,
+        cache = cache_put(Table, IntK, V, Cache),
+        dirty = Dirty#{IntK => dirty}
+    }.
+
+get_or_gen_internal_key(K, KeyMapping, _Domain, _Cache) when
+    is_map_key(K, KeyMapping)
+->
+    IntK = maps:get(K, KeyMapping),
+    {IntK, KeyMapping};
+get_or_gen_internal_key(K, KeyMapping0, Domain, Cache) ->
+    IntK = erlang:unique_integer(),
```

Comment:
Must generate keys in a different manner for streams to preserve their order relation.